### PR TITLE
Add ByteArray constructors (#3879) 

### PR DESCRIPTION
**Pull request by tustvold**
# Which issue does this PR close?

Part of #3879

# Rationale for this change
 

# What changes are included in this PR?

# Are there any user-facing changes?

### CODE REVIEW

File: arrow-array/src/types.rs [hunk 1]
```diff
@@ -1569,6 +1604,21 @@ impl<O: OffsetSizeTrait> ByteArrayType for GenericBinaryType<O> {
...
+    fn validate(
+        offsets: &OffsetBuffer<Self::Offset>,
+        values: &Buffer,
+    ) -> Result<(), ArrowError> {
+        // offsets are guaranteed to be monotonically increasing and non-empty
+        let max_offset = offsets.last().unwrap().as_usize();
+        if values.len() < max_offset {
+            return Err(ArrowError::InvalidArgumentError(format!(
+                "Maximum offset of {max_offset} is larger than values of length {}",
+                values.len()
```
Comment (viirya) [inline arrow-array/src/types.rs:1617]:
```suggestion
                "Maximum offset of {max_offset} is larger than length of values {}",
                values.len()
```

File: arrow-array/src/array/byte_array.rs [hunk 2]
```diff
@@ -60,6 +60,85 @@ impl<T: ByteArrayType> GenericByteArray<T> {
...
+    pub fn try_new(
+        offsets: OffsetBuffer<T::Offset>,
+        values: Buffer,
+        nulls: Option<NullBuffer>,
+    ) -> Result<Self, ArrowError> {
+        let len = offsets.len() - 1;
+        T::validate(&offsets, &values)?;
+
+        if let Some(n) = nulls.as_ref() {
+            if n.len() != len {
```
Comment (viirya) [inline arrow-array/src/array/byte_array.rs:91]: Don't we need to also check the length of values buffer?

**Review by viirya — APPROVED** [review]

File: arrow-array/src/types.rs (hunk 1 above)
Comment (tustvold) [reply arrow-array/src/types.rs:1617]: I think the phrasing before is grammatically more correct

File: arrow-array/src/array/byte_array.rs (hunk 2 above)
Comment (tustvold) [reply arrow-array/src/array/byte_array.rs:91]: This is covered by `T::validate`, will add a comment

**Pull request merged**